[py] Now the `ClientWindowInfo` object's attributes are validated through `getters/setters`

sandeepsuryaprasad opened the pull request:
### **User description**
<!-- Thanks for Contributing to Selenium! -->
<!-- Please read our contribution guidelines: https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md -->

### 🔗 Related Issues
<!-- Example: Fixes #1234 or Closes #5678 -->
<!-- If the reason for this PR is not obvious, consider creating an issue for it first -->

### 💥 What does this PR do?
In this PR, I have encapsulated all instance attributes of `ClientWindowInfo` class in `getters/setters` and now the attributes are being validated through `setters`

### 🔧 Implementation Notes
It is not a standard practice to validate instance attributes inside a class method. More Pythonic approach is to have separate `getters/setters` to do the validations. The class method `from_dict` is cluttered with validations of different instance attributes before the class is getting instantiated. The more standard and structured approach for this is to validate through `setters`. After the changes made in this PR, the class method looks clean and readable.

In the current implementation although there are few getter methods like, `get_state`, `get_client_window`, `get_width`, `get_height`, `get_x`, `get_y`, which returns the value of corresponding attributes, the setting of these attributes are done in constructor, but the validations before they are being set is done in class method `from_dict` which I felt is bit strange. 

The getters in the current code is not in pythonic style and it is Java way of doing things. The more pythonic approach is to have `property` descriptors, which I have implemented in this PR.

### 💡 Additional Considerations
<!--- Are there any decisions that need to be made before accepting this PR? -->
<!--- Is there any follow-on work that needs to be done? (e.g., docs, tests, etc.) -->

### 🔄 Types of changes
<!-- ✂️ Please delete anything that doesn't apply -->
- Cleanup (formatting, renaming)


___

### **PR Type**
Enhancement


___

### **Description**
- Refactored `ClientWindowInfo` class to use Python properties instead of getter methods

- Added validation in property setters for all attributes

- Simplified `from_dict` class method by removing inline validations

- Updated tests to use property access instead of getter methods


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Old getter methods"] --> B["Python properties"]
  C["from_dict validations"] --> D["Property setter validations"]
  E["Test getter calls"] --> F["Direct property access"]
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>browser.py</strong><dd><code>Convert getters to properties with validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/common/bidi/browser.py

<ul><li>Replaced getter methods with Python <code>@property</code> decorators<br> <li> Added setter methods with validation logic for all attributes<br> <li> Simplified <code>from_dict</code> method by removing validation code<br> <li> Added comprehensive docstrings for all properties</ul>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16093/files#diff-cd7018a350739e3db649c9e7f186036670f27a06f8df204de5909a03d7a04f5f">+109/-54</a></td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>bidi_browser_tests.py</strong><dd><code>Update tests for property access</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/test/selenium/webdriver/common/bidi_browser_tests.py

<ul><li>Updated test assertions to use property access instead of getter <br>methods<br> <li> Changed <code>get_client_window()</code> to <code>client_window</code> property access<br> <li> Changed <code>get_state()</code>, <code>get_width()</code>, <code>get_height()</code>, <code>get_x()</code>, <code>get_y()</code> to <br>direct property access</ul>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16093/files#diff-e0a8b792e4fc6fa62c5fae3b9c75b56d61e12e4d0929aa991ddb799276b53c3c">+7/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

sandeepsuryaprasad committed:
moved all instance attribute validations to setters in ClientWindowInfo object

sandeepsuryaprasad committed:
fixed typo

sandeepsuryaprasad committed:
fixed typo

sandeepsuryaprasad committed:
added doc strings

sandeepsuryaprasad committed:
clean up

sandeepsuryaprasad committed:
modified doc string of class method

sandeepsuryaprasad committed:
fixed formatting

navin772 committed:
Merge branch 'trunk' into browser

sandeepsuryaprasad committed:
fixed formatting issue

sandeepsuryaprasad commented:
@navin772 can you please re-trigger the CI. I have fixed the formatting issue that i missed.. Thanks!

sandeepsuryaprasad commented:
@cgoldberg can we have this one merged..

sandeepsuryaprasad commented:
@cgoldberg @navin772 can we have this is one merged.. I think the test that got failed is not related to the change made in this PR.

sandeepsuryaprasad commented:
@cgoldberg @shbenzer can we have this PR merged..

navin772 commented on py/test/selenium/webdriver/common/bidi_browser_tests.py:
```diff
@@ -99,14 +99,14 @@ def test_get_client_windows(driver):
-    assert window_info.get_height() > 0
+    assert window_info.client_window is not None
+    assert window_info.state is not None
+    assert isinstance(window_info.state, str)
+    assert window_info.width > 0
+    assert window_info.height > 0
     assert isinstance(window_info.is_active(), bool)
-    assert window_info.get_x() >= 0
-    assert window_info.get_y() >= 0
+    assert window_info.x >= 0
+    assert window_info.y >= 0
```
The changes look good, but we are essentially changing the method/property names which is a breaking change. It's not that the new property names are bad, they are better and more intuitive. 

The BiDi browser module is relatively new but I am still not sure if we do this change without notification. What do you think @cgoldberg?

cgoldberg commented:
> changing the method/property names which is a breaking change

yes, this PR breaks the API. I don't want to merge it as-is and break any existing code out there.

I think the only way we can accept these changes is to keep the old methods functional and add deprecation warnings, along with some `TODO` comments that they should be removed later. (I think our deprecation policy is to keep them for 3 releases?)

cgoldberg reviewed (changes requested):
see previous comments about deprecating old methods.

sandeepsuryaprasad committed:
added deprecations

sandeepsuryaprasad committed:
reverted old getter methods in unit tests

sandeepsuryaprasad commented:
@cgoldberg @navin772 I have added all `get_` methods back along with deprecation warnings. Let me know if it is okay.

navin772 reviewed (changes requested):
You should add test for the newly added property not the deprecated ones.
The `get_client_window()` method is just a wrapper around `client_window` property.

sandeepsuryaprasad committed:
replaced get_ methods with property in unittest

sandeepsuryaprasad commented:
@navin772 @cgoldberg I have replaced unit test with  corresponding properties.

navin772 commented:
@sandeepsuryaprasad could you run the linter, looks like `Any` import from typing is missing

sandeepsuryaprasad commented:
@navin772 ohh yes.. i have added it now.. please check..

navin772 commented on py/selenium/webdriver/common/bidi/browser.py:
```diff
@@ -62,53 +63,207 @@ def get_state(self) -> str:
+            raise ValueError(f"y must be an integer, got {type(value).__name__}")
+        self._y = value
+
+    @property
+    def active(self):
+        """Gets the Window Status.
+
+        Returns:
+        -------
+            bool: The boolen value of Window Status
+        """
+        return self._active
+
+    @active.setter
+    def active(self, value) -> None:
+        """Sets the Window Status.
+
+        Returns: None
+        """
+        if not isinstance(value, bool):
+            raise ValueError("active must be a boolean")
+        self._active = value
```
We need to deprecate the `is_active()` method too. Also, add the return type for `active` property:
```python
def active(self) -> bool:
```

navin772 commented:
This PR introduces new mypy type errors for the `bidi/browser.py` file:

```
selenium/webdriver/common/bidi/browser.py:289: error: Argument "client_window" to "ClientWindowInfo" has incompatible type "Any | None"; expected "str"  [arg-type]
selenium/webdriver/common/bidi/browser.py:290: error: Argument "state" to "ClientWindowInfo" has incompatible type "Any | None"; expected "str"  [arg-type]
selenium/webdriver/common/bidi/browser.py:291: error: Argument "width" to "ClientWindowInfo" has incompatible type "Any | None"; expected "int"  [arg-type]
selenium/webdriver/common/bidi/browser.py:292: error: Argument "height" to "ClientWindowInfo" has incompatible type "Any | None"; expected "int"  [arg-type]
selenium/webdriver/common/bidi/browser.py:293: error: Argument "x" to "ClientWindowInfo" has incompatible type "Any | None"; expected "int"  [arg-type]
selenium/webdriver/common/bidi/browser.py:294: error: Argument "y" to "ClientWindowInfo" has incompatible type "Any | None"; expected "int"  [arg-type]
selenium/webdriver/common/bidi/browser.py:295: error: Argument "active" to "ClientWindowInfo" has incompatible type "Any | None"; expected "bool"  [arg-type]
```

@sandeepsuryaprasad could you please fix them, we are trying to zero down the mypy errors and enable them in CI checks.

sandeepsuryaprasad committed:
fixed mypy errors

sandeepsuryaprasad committed:
Merge branch 'trunk' into browser

sandeepsuryaprasad commented:
@navin772 I have fixed mypy errors.. mypy should be happy now.. Also, I have added deprecation warning to `is_active` method and added return type for `active` property

navin772 committed:
Merge branch 'trunk' into browser

navin772 commented on py/test/selenium/webdriver/common/bidi_browser_tests.py:
```diff
@@ -99,14 +99,14 @@ def test_get_client_windows(driver):
+    assert isinstance(window_info.state, str)
+    assert window_info.width > 0
+    assert window_info.height > 0
     assert isinstance(window_info.is_active(), bool)
```
Use `active` property here instead of deprecated `is_active()`

sandeepsuryaprasad replied:
Done..

sandeepsuryaprasad committed:
Update bidi_browser_tests.py

navin772 committed:
Merge branch 'trunk' into browser

navin772 reviewed (approved):
LGTM!

sandeepsuryaprasad commented:
@cgoldberg can you please give your final review comments on this PR. Thanks!

sandeepsuryaprasad commented:
@shbenzer @cgoldberg can you please review this PR.. it's long over-due for your review. Thanks!

sandeepsuryaprasad committed:
Merge branch 'trunk' into browser

Delta456 committed:
Merge branch 'trunk' into browser

sandeepsuryaprasad committed:
fixed linting
